### PR TITLE
Add hidden players and objects.

### PR DESCRIPTION
**Pull request by luk3yx**
Allows players and other objects to be hidden from clients with `object:hide()`.
For #8316 

Known issues:
* ~~Hiding players as soon as they join prevents `F7` from working until they are unhidden.~~ Fixed
* ~~Hidden player names are sent to clients.~~ Fixed
* The fake add/remove player messages sent to clients seem hacky, maybe there's a better way.
* The added documentation probably needs changing.

### CODE REVIEW

**Comment by bigfoot547** [comment]
Wow, I'd love to see this implemented into the game! Good code too :)

**Comment by paramat** [comment]
> * `hide()`: Hides the object. *Hidden objects cannot be seen by players.*

Sorry for all the questions but there is no explanation.
Hidden from all players? What do you mean by 'hidden'? Just visually invisible but collidable? Can this not be done another way ising node properties? What's the use case or need? Is there a related issue?

**Comment by luk3yx** [comment]
> > * `hide()`: Hides the object. _Hidden objects cannot be seen by players._
> 
> Sorry for all the questions but there is no explanation.
> Hidden from all players? What do you mean by 'hidden', just visually invisible but collidable? Can this not be done another way using object properties?
> 
> Issue #8316

Hidden objects are not sent to clients (and therefore can't be collided with or be displayed on minimaps).

**Comment by paramat** [comment]
Sorry, i deleted my post because the issue explained.

File: src/clientiface.cpp [hunk 1]
```diff
@@ -699,11 +699,15 @@ void ClientInterface::UpdatePlayerList()
 		for (session_t i : clients) {
 			RemotePlayer *player = m_env->getPlayer(i);
 
-			if (player == NULL)
+			if (player == nullptr)
```
Comment (ClobberXD) [inline src/clientiface.cpp]:
```suggestion
			if (!player)
```

File: src/clientiface.cpp [hunk 2]
```diff
@@ -699,11 +699,15 @@ void ClientInterface::UpdatePlayerList()
...
 				continue;
 
 			infostream << "* " << player->getName() << "\t";
 
+			PlayerSAO *sao = player->getPlayerSAO();
+			if (sao != nullptr && sao->isHidden())
```
Comment (ClobberXD) [inline src/clientiface.cpp]:
```suggestion
			if (sao && sao->isHidden())
```

File: src/content_sao.cpp [hunk 3]
```diff
@@ -725,6 +719,21 @@ u16 LuaEntitySAO::getHP() const
...
+{
+	m_hidden = true;
+}
+
+void LuaEntitySAO::unhide()
+{
+	m_hidden = false;
+}
+
+bool LuaEntitySAO::isHidden() const
```
Comment (ClobberXD) [inline src/content_sao.cpp]:
```suggestion
const bool LuaEntitySAO::isHidden() const
```

File: src/content_sao.h [hunk 4]
```diff
@@ -129,6 +129,9 @@ class LuaEntitySAO : public UnitSAO
 	std::string getDescription();
 	void setHP(s32 hp, const PlayerHPChangeReason &reason);
 	u16 getHP() const;
+	void hide();
+	void unhide();
+	bool isHidden() const;
```
Comment (ClobberXD) [inline src/content_sao.h]:
```suggestion
	const bool isHidden() const;
```

File: src/script/lua_api/l_object.cpp [hunk 5]
```diff
@@ -761,6 +761,53 @@ int ObjectRef::l_get_properties(lua_State *L)
...
 }
 
+// hide(self)
+int ObjectRef::l_hide(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	ServerActiveObject *co = getobject(ref);
+
+	if (co == NULL)
```
Comment (ClobberXD) [inline src/script/lua_api/l_object.cpp]:
```suggestion
	if (!co)
```

File: src/script/lua_api/l_object.cpp [hunk 6]
```diff
@@ -761,6 +761,53 @@ int ObjectRef::l_get_properties(lua_State *L)
...
+}
+
+// unhide(self)
+int ObjectRef::l_unhide(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	ServerActiveObject *co = getobject(ref);
+
+	if (co == NULL)
```
Comment (ClobberXD) [inline src/script/lua_api/l_object.cpp]:
```suggestion
	if (!co)
```

File: src/script/lua_api/l_object.cpp [hunk 7]
```diff
@@ -761,6 +761,53 @@ int ObjectRef::l_get_properties(lua_State *L)
...
+}
+
+// is_hidden(self)
+int ObjectRef::l_is_hidden(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	ServerActiveObject *co = getobject(ref);
+
+	if (co == NULL)
```
Comment (ClobberXD) [inline src/script/lua_api/l_object.cpp]:
```suggestion
	if (!co)
```

File: src/serverobject.h [hunk 8]
```diff
@@ -145,6 +145,10 @@ class ServerActiveObject : public ActiveObject
 	virtual u16 getHP() const
 	{ return 0; }
 
+	virtual void hide() { m_hidden = true; }
+	virtual void unhide() { m_hidden = false; }
+	virtual bool isHidden() { return m_hidden; }
```
Comment (ClobberXD) [inline src/serverobject.h]:
```suggestion
	virtual const bool isHidden() { return m_hidden; }
```

**Comment by stujones11** [comment]
Seems like it might be useful, though I would prefer something like `set_visible(<bool>)` and `is_visible()` but it's not that big a deal I guess :)

**Comment by benrob0329** [comment]
Allowing mods to hide objects on a per-player basis would add to the usefulness this provides.

**Comment by stujones11** [comment]
> Allowing mods to hide objects on a per-player basis would add to the usefulness this provides.

It would also add to the complexity so may be better made as a separate PR if we want to see it merged in our own lifetimes :P

Though I do fully support that idea.

**Comment by ClobberXD** [comment]
I also agree that it should be done as a separate PR, for the same reason ;)

**Comment by paramat** [comment]
> It would also add to the complexity so may be better made as a separate PR

Agreed. Personally i'm neutral about this PR but am not convinced hiding on a per-player basis is justified, please discuss that in an issue before making any PR.

**Comment by luk3yx** [comment]
Is there anything blocking this from being merged (apart from core dev approval)?

File: src/content_sao.cpp [hunk 9]
```diff
@@ -1349,6 +1358,38 @@ void PlayerSAO::setBreath(const u16 breath, bool send)
 		m_env->getGameDef()->SendPlayerBreath(this);
 }
 
+void PlayerSAO::hide()
+{
+	if (m_hidden)
+		return;
+
+	m_hidden = true;
+	if (! m_player)
```
Comment (ClobberXD) [inline src/content_sao.cpp]:
```suggestion
	if (!m_player)
```

Please also remove spaces between `!` and variable name everywhere else within the changes too.

**Comment by SmallJoker** [comment]
Is there any reason to not integrate this as an object property? That way you could put it as a value into the entity init packet which is sent to the players, so that the players won't even add them once to the object list.

**Comment by luk3yx** [comment]
I'd rather not send hidden objects to clients at all if possible.

File: src/clientiface.cpp [hunk 10]
```diff
@@ -699,7 +699,11 @@ void ClientInterface::UpdatePlayerList()
 		for (session_t i : clients) {
 			RemotePlayer *player = m_env->getPlayer(i);
 
-			if (player == NULL)
+			if (!player)
+				continue;
+
+			PlayerSAO *sao = player->getPlayerSAO();
+			if (sao && sao->isHidden())
```
Comment (nerzhul) [inline src/clientiface.cpp]: sorry but SAO has nothing to do with Player list. it's not responsibility of an AO property to update player list

File: src/clientiface.h [hunk 11]
```diff
@@ -467,6 +467,8 @@ class ClientInterface {
 	/* get state of client by id*/
 	ClientState getClientState(session_t peer_id);
 
+	void sendFakeJoinLeaveMessage(std::string player_name, bool in_game);
```
Comment (nerzhul) [inline src/clientiface.h]: const std::string &

File: src/content_sao.cpp [hunk 12]
```diff
@@ -725,6 +719,21 @@ u16 LuaEntitySAO::getHP() const
 	return m_hp;
 }
 
+void LuaEntitySAO::hide()
```
Comment (nerzhul) [inline src/content_sao.cpp]: inline those functions please. And change to hide(bool), we don't need 2 functions

File: src/content_sao.cpp [hunk 13]
```diff
@@ -1349,6 +1358,38 @@ void PlayerSAO::setBreath(const u16 breath, bool send)
...
+
+	m_hidden = true;
+	if (!m_player)
+		return;
+
+	Server *server = m_env->getGameDef();
+	if (!server)
+		return;
+
+	server->sendFakeJoinLeaveMessage(m_player->getName(), false);
```
Comment (nerzhul) [inline src/content_sao.cpp]: sounds very hacky to me

**Comment by SmallJoker** [comment]
@luk3yx In that case, the LuaEntitySAO/PlayerSAO init function can be skipped in the server code if that property is set.

File: src/content_sao.cpp (hunk 13 above)
Comment (luk3yx) [reply src/content_sao.cpp]: I agree (and mentioned this in my initial comment), do you or anyone else have a better and less hacky idea?

File: src/content_sao.cpp (hunk 12 above)
Comment (luk3yx) [reply src/content_sao.cpp]: Should I change the lua API as well?

File: src/clientiface.cpp (hunk 10 above)
Comment (luk3yx) [reply src/clientiface.cpp]: Do you have a better idea?

File: src/content_sao.cpp (hunk 13 above)
Comment (ClobberXD) [reply src/content_sao.cpp]: Would it be less hacky to directly send a chat message?

**Comment by BluebirdGreycoat** [comment]
Any thing going on with this?

It's kinda hard to check if someone's hacking if they can see you coming, first.

**Comment by luk3yx** [comment]
> Any thing going on with this?
> 
> It's kinda hard to check if someone's hacking if they can see you coming, first.

This PR needs an API change (`:hide()` and `:unhide()` to `:set_visible(<bool>)`), a less hacky way of sending fake join/leave packets, and possibly a `hidden` attribute on players as well as player SAOs.

If you don't care about invisibility from chat tab-complete and minimaps, there are a few invisibility mods around:

 - [cloaking](https://forum.minetest.net/viewtopic.php?t=20788): Attempts to hide players from both clients and mods (unless mods use cloaking's API).
 - [incognito](https://forum.minetest.net/viewtopic.php?t=19050): Similar, however players aren't hidden from mods.
 - [invis](https://forum.minetest.net/viewtopic.php?t=17389): A simple-ish mod that makes players invisible without sending fake join/leave messages or hiding players from other mods.

*Note that the `incognito` and `invis` mods may not work with Minetest 5.0.0+, a workaround would be to change reference(s) to `collisionbox` with `selectionbox`.*

**Comment by paramat** [comment]
Author seems to need help from others, might need adoption so added the label.

**Pull request closed without merging**